[Concept Entry] Ruby: JRuby (#3006)
Codecademy/docs: content/ruby/concepts/jruby/jruby.md

---
Title: 'JRuby'
Description: 'JRuby is an implementation of Ruby written in Java programming language using the Java Virtual Machine (JVM).'
Subjects:
  - 'Web Development'
  - 'Web Design'
Tags:
  - 'JRuby'
CatalogContent:
  - 'learn-ruby'
  - 'learn-rails'
---

> 

Reviewer: ```suggestion
There should be an introductory paragraph without a header. It can be the same as the Metadata description.
```

Author: I performed this task in CLI. I will send another PR and link it to #3006

> ## Design Differences

Reviewer: ```suggestion
## Design Differences

```
Should be a blank line after each header.

Ruby is written and implemented in the C programming language, and JRuby is written and implemented in Java. Ruby and JRuby use object-oriented programming, dynamic typing, and garbage collection despite being written in different programming languages. As a result, JRuby can understand and interpret all ruby code with a few exceptions. 

> ## Interpreter differences

Reviewer: ```suggestion
## Interpreter differences

```

JRuby will interpret ruby code using Java's Virtual Machine (JVM) rather than Ruby's interpreter. All of Ruby's built-in core classes are supported by JRuby and most standard libraries; however, there are a few missing libraries that depend on a C language implementation. 

> ### Global Interrupt Lock

Reviewer: ```suggestion
### Global Interrupt Lock

```

> While Ruby can create threads and perform serial execution, Ruby has a Global Interrupter Lock, which means Ruby can only execute one code at a time. JRuby does not have this limitation and uses the JVM multi-threaded features to complete tasks faster by running code simultaneously.

Reviewer: ```suggestion
While Ruby can create threads and perform serial execution, Ruby has a Global Interrupter Lock, which means Ruby can only execute one thread at a time. JRuby does not have this limitation and uses the JVM's multi-threaded features to complete tasks faster by running code simultaneously.
```

> ### JRuby Advantages

Reviewer: ```suggestion
### JRuby Advantages

```

> The advantages of using JRuby are high performance, real threading, and platform independence. JRuby achieves more requests per second by using the JVM's multi-threading. The JVM has over 20 years of experienced engineers working to solve many performance issues by enhancing concurrency, parallelism, memory garbage collection, and Just-in-Time(JIT) compiler. 

Reviewer: ```suggestion
The advantages of using JRuby are high performance, real threading, and platform independence. JRuby achieves more requests per second by using the JVM's multi-threading capability. The JVM has over 20 years of experienced engineers working to solve many performance issues by enhancing concurrency, parallelism, memory garbage collection, and the Just-in-Time(JIT) compiler. 
```

> ### JRuby Disadvantages

Reviewer: ```suggestion
### JRuby Disadvantages

```

JRuby takes longer to start up, tends to be more memory intensive, and may even overwrite memory which could result in additional time spent addressing bugs. JRuby may lose access to the C library compatibility, and some gems may not transition to JRuby because of incompatible native C libraries. 

> ## Some popular Java Ruby implementations:

Reviewer: ```suggestion
## Some popular Java Ruby implementations

```

---
Title: 'JRuby'
Description: 'JRuby is an implementation of Ruby written in Java programming language using the Java Virtual Machine (JVM).'
Subjects:
  - 'Web Development'
  - 'Web Design'
Tags:
  - 'JRuby'
CatalogContent:
  - 'learn-ruby'
  - 'learn-rails'
---

> JRuby is an implementation of Ruby written in Java programming language using the Java Virtual Machine (JVM).

Reviewer 2: ```suggestion
JRuby is an implementation of Ruby written in Java that leverages the Java Virtual Machine (JVM), which provides performance and accessibility benefits for deploying Ruby applications.
```
